plotter/sankey.go

```diff
@@ -0,0 +1,440 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package plotter
+
+import (
+	"fmt"
+	"image/color"
+	"math"
+	"sort"
+
+	"github.com/gonum/floats"
+	"github.com/gonum/plot"
+	"github.com/gonum/plot/vg"
+	"github.com/gonum/plot/vg/draw"
+	"github.com/ready-steady/spline"
+)
+
+// A Sankey diagram presents stock and flow data as rectangles representing
+// the amount of each stock and lines between the stocks representing the
+// amount of each flow.
+type Sankey struct {
+	// StockPad specifies the padding between
+	// stocks in the same category, in chart units.
+	StockPad float64
+
+	// Color specifies the default fill
+	// colors for the stocks and flows. Colors can be
+	// modified for individual stocks and flows.
+	color.Color
+
+	// StockBarWidth is the widths of the bars representing
+	// the stocks. The default value is 15% larger than the
+	// height of the stock label text.
+	StockBarWidth vg.Length
+
+	// LineStyle specifies the default border
+	// line style for the stocks and flows. Styles can be
+	// modified for individual stocks and flows.
+	draw.LineStyle
+
+	// TextStyle specifies the default stock label
+	// text style. Styles can be modified for
+	// individual stocks.
+	draw.TextStyle
+
+	flows []Flow
+
+	// FlowStyle is a function that specifies the
+	// background colar and border line style of the
+	// flow based on its group name. The default
+	// function uses the default Color and LineStyle
+	// specified above for all groups.
+	FlowStyle func(group string) (color.Color, draw.LineStyle)
+
+	// stocks arranges the stocks by category.
+	// The first key is the category and the seond
+	// key is the label.
+	stocks map[int]map[string]*stock
+}
+
+// stock represents the amount of a stock and its plotting order.
+type stock struct {
+	receptorValue, sourceValue float64
+	label                      string
+	category                   int
+	order                      int
+
+	// min represents the beginning of the plotting location
+	// on the value axis.
+	min float64
+
+	// max is min plus the larger of receptorValue and sourceValue.
+	max float64
+
+	// sourceFlowPlaceholder and receptorFlowPlaceholder track
+	//  the current plotting location during
+	// the plotting process.
+	sourceFlowPlaceholder, receptorFlowPlaceholder float64
+}
+
+// A Flow represents the amount of an entity flowing between two stocks.
+type Flow struct {
+	// SourceStockLabel and ReceptorStockLabel are the labels
+	// of the stocks that originate and receive the flow,
+	// respectively.
+	SourceStockLabel, ReceptorStockLabel string
```

Comment:
Fair enough, the issue for me is more that it makes the field names java-esque. If Source and Destination are meaningful and short. SourceCategory and DestinationCategory are unfortunately longer, but an additional type to allay this doesn't seem warranted.